serverless/serverless
# Add Application Load Balancer event source

## What did you implement:

Closes #5572 

Adds support for the `alb` event type.

---

**Shoutouts** to @cbm-egoubely who provided his plugin (see comments below) and the design for the abstraction we're using here. 💯 🥇

---

## How did you implement it:

Added all the code to get an `alb` event setup up and running.

## How can we verify it:

Configuration is based on [this `serverless.yml`](https://github.com/jthomerson/serverless-training-examples/blob/master/services/application-load-balancer-lambda/serverless.yml) and [this blog post](https://serverless-training.com/articles/how-to-set-up-application-load-balancer-with-lambda/).

Use the following `serverless.yml` and run `serverless deploy -v`:

Notice that the `LoadBalancerDNSName` output is printed on the terminal. Copy this value and append the `/hello` path/. `curl` it.

```javascript
// handler.js

module.exports.hello = async () => ({
  isBase64Encoded: false,
  statusCode: 200,
  statusDescription: "200 OK",
  headers: { "Set-cookie": "cookies", "Content-Type": "application/json" },
  body: "Hello from Lambda"
});
```

```yaml
service: test

provider:
  name: aws
  runtime: nodejs10.x
  versionFunctions: false

functions:
  test:
    handler: handler.hello
    events:
      - alb:
          listenerArn: { Ref: HTTPListener }
          priority: 1
          conditions:
            path: /hello

custom:
  defaultRegion: us-east-1
  defaultStage: dev
  region: ${opt:region, self:custom.defaultRegion}
  stage: ${opt:stage, self:custom.defaultStage}
  objectPrefix: '${self:service}-${self:custom.stage}'

resources:
  Outputs:
    LoadBalancerDNSName:
      Value: { 'Fn::GetAtt': [ LoadBalancer, 'DNSName' ] }
      Export: { Name: '${self:custom.objectPrefix}-LoadBalancerDNSName' }
  Resources:
    VPC:
      Type: 'AWS::EC2::VPC'
      Properties:
        CidrBlock: 172.31.0.0/16
        EnableDnsHostnames: true
    InternetGateway:
      Type: 'AWS::EC2::InternetGateway'
    VPCGatewayAttachment:
      Type: 'AWS::EC2::VPCGatewayAttachment'
      Properties:
        VpcId: { Ref: VPC }
        InternetGatewayId: { Ref: InternetGateway }
    RouteTable:
      Type: 'AWS::EC2::RouteTable'
      Properties:
        VpcId: { Ref: VPC }
    InternetRoute:
      Type: 'AWS::EC2::Route'
      DependsOn: VPCGatewayAttachment
      Properties:
        DestinationCidrBlock: 0.0.0.0/0
        GatewayId: { Ref: InternetGateway }
        RouteTableId: { Ref: RouteTable }
    SubnetA:
      Type: 'AWS::EC2::Subnet'
      Properties:
        AvailabilityZone: '${self:custom.region}a'
        CidrBlock: 172.31.0.0/20
        MapPublicIpOnLaunch: false
        VpcId: { Ref: VPC }
    SubnetB:
      Type: 'AWS::EC2::Subnet'
      Properties:
        AvailabilityZone: '${self:custom.region}b'
        CidrBlock: 172.31.16.0/20
        MapPublicIpOnLaunch: false
        VpcId: { Ref: VPC }
    SubnetARouteTableAssociation:
      Type: 'AWS::EC2::SubnetRouteTableAssociation'
      Properties:
        SubnetId: { Ref: SubnetA }
        RouteTableId: { Ref: RouteTable }
    SubnetBRouteTableAssociation:
      Type: 'AWS::EC2::SubnetRouteTableAssociation'
      Properties:
        SubnetId: { Ref: SubnetB }
        RouteTableId: { Ref: RouteTable }
    SecurityGroup:
      Type: 'AWS::EC2::SecurityGroup'
      Properties:
        GroupName: 'http-https'
        GroupDescription: 'HTTPS / HTTPS inbound; Nothing outbound'
        VpcId: { Ref: VPC }
        SecurityGroupIngress:
          -
            IpProtocol: tcp
            FromPort: '80'
            ToPort: '80'
            CidrIp: 0.0.0.0/0
          -
            IpProtocol: tcp
            FromPort: '443'
            ToPort: '443'
            CidrIp: 0.0.0.0/0
        SecurityGroupEgress:
          -
            IpProtocol: -1
            FromPort: '1'
            ToPort: '1'
            CidrIp: 127.0.0.1/32
    LoadBalancer:
      Type: 'AWS::ElasticLoadBalancingV2::LoadBalancer'
      Properties:
        Type: 'application'
        Name: '${self:custom.objectPrefix}'
        IpAddressType: 'ipv4'
        Scheme: 'internet-facing'
        LoadBalancerAttributes:
          - { Key: 'deletion_protection.enabled', Value: false }
          - { Key: 'routing.http2.enabled', Value: false }
          - { Key: 'access_logs.s3.enabled', Value: false }
        SecurityGroups:
          - { Ref: SecurityGroup }
        Subnets:
          - { Ref: SubnetA }
          - { Ref: SubnetB }
    HTTPListener:
      Type: 'AWS::ElasticLoadBalancingV2::Listener'
      Properties:
        LoadBalancerArn: { Ref: LoadBalancer }
        Port: 80
        Protocol: 'HTTP'
        DefaultActions:
          -
            Type: 'fixed-response'
            Order: 1
            FixedResponseConfig:
              StatusCode: 404
              ContentType: 'application/json'
              MessageBody: '{ "not": "found" }'
```

## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
